lib/std/os/windows.zig

```diff
@@ -523,19 +527,186 @@ pub fn ReadFile(in_hFile: HANDLE, buffer: []u8, offset: ?u64, io_mode: std.io.Mo
                 };
                 break :blk &overlapped_data;
             } else null;
-            if (kernel32.ReadFile(in_hFile, buffer.ptr, want_read_count, &amt_read, overlapped) == 0) {
+            var console_mode: DWORD = undefined;
+            const is_console_handle: bool = kernel32.GetConsoleMode(in_hFile, &console_mode) != FALSE;
+            const is_cooked_mode: bool = (console_mode & ENABLE_LINE_INPUT) != 0;
+            // Implementation issue:
+            // There is no reliable way to implement perfectly platform-agnostic UTF-16 to UTF-8
+            // conversion for raw mode, because it is impossible to know the number of pending
+            // code units stored in console input buffer, while in cooked mode we can rely on the
+            // terminating LF character. Without knowing that, ReadConsoleW() may accidentally pop
+            // out characters without blocking, or prompt for user input at unexpected timing.
+            // In the case of raw mode, redirect to kernel32.ReadFile() without conversion for now,
+            // just don't make things worse.
```

Comment:
This is not good enough. If it cannot be implemented perfectly using the current abstractions, then the current abstractions need to change.

Reply:
Let me elaborate on this:

The current goal of my implementation is to make users totally unaware of UTF-16 presence even if they call `std.os.windows.ReadFile` directly. Users are able to pass UTF-8-encoded strings everywhere, and `std.os.windows.ReadFile` does the translation job under the hood. It means that I need to insert a UTF-16 → UTF-8 translation layer between `std.os.windows.ReadFile` and `kernel32.ReadFile`.

Here comes the key problem that gives rise to annoying edge cases: The user-provided buffer is meant to be UTF-8 encoded. Console input (which is UTF-16 encoded) must be converted to UTF-8 before filling into the user-provided buffer, while the converted UTF-8 length can't be easily determined. When the user-provided buffer is not large enough to hold the converted console input, there are three pieces of data we have to deal with carefully:

* UTF-8 fragments. For example, a Unicode character from console input converts into a 3-byte sequence, but the user-provided buffer is only 2 bytes long. Then `ReadFile` should store the first 2 bytes into the user-provided buffer and stash the final byte. Whenever `ReadFile` is called again, this byte should be popped first before reading anything else from the console.
* Single UTF-16 high surrogate yet to form a complete Unicode code point. We should stash it separately and wait for a low surrogate from another `ReadConsoleW` call to convert into UTF-8 byte sequence.
* UTF-16 code units left in `ReadConsoleW`'s dedicated buffer. It is completely opaque to user-mode code, causing serious troubles, which will be discussed below.

There are two modes affecting `ReadConsoleW` behaviour determined by `ENABLE_LINE_INPUT` flag (accessed with `GetConsoleMode`/`SetConsoleMode` functions):

* Cooked mode (`ENABLE_LINE_INPUT` set): If CR is encountered in the dedicated buffer, an LF is appended to the user-provided buffer and reading from console ends immediately. Otherwise, wait for console input until a CR is entered (either single key press or pasting multiple characters). Everything after the CR is left inside the dedicated buffer after the read.
* Raw mode (`ENABLE_LINE_INPUT` unset): If there is at least one character available in the dedicated buffer, reading from console ends immediately. Otherwise, wait for console input (either single key press or pasting multiple characters). The dedicated buffer is flushed after the read.

Cooked mode can be emulated perfectly, as in my implementation. However, to emulate the behaviour in raw mode, we have to try flushing UTF-8 fragments and then call `ReadConsoleW` to flush the dedicated buffer. The problem rises when we popped at least one byte from the UTF-8 fragments. Because we have "read" something (from UTF-8 fragments), `std.os.windows.ReadFile` should return immediately if the `ReadConsoleW` dedicated buffer is empty, but calling `ReadConsoleW` when the dedicated buffer is empty causes an unwanted wait for console input. As I explained before, there is no way to know whether the dedicated buffer is empty other than calling `ReadConsoleW`, so it is a stalemate.

If there are better abstractions, please let me know.

Reply:
Relevant commit message from Go: https://github.com/golang/go/commit/610d522189ed3fcf0d298609a248a3283bde62cd

The `readConsole` implementation has changed a bit since then, the latest is here: https://github.com/golang/go/blob/6e8caefc19cae465444775f6cd107b138a26cce7/src/internal/poll/fd_windows.go#L454-L521

There does not seem to be any special handling of `ENABLE_LINE_INPUT` in Rust or Go at all.